### DISCUSSION TITLE
C/C++ implementation of the model

### DISCUSSION THREAD

**Discussion by ggerganov** [openai/whisper#130] [Show and tell]
Great work on this! The whisper model is very interesting and I think is going to open a lot of interesting possibilities in the future (not just for speech-to-text).

This weekend, as a learning exercise, I decided to implement the inference from scratch in C/C++:

https://github.com/ggerganov/whisper.cpp

I'm happy with the results. I'm still struggling with the beam search sampling strategy that is used in the original model. Still, the greedy implementation that I hacked seems to do the job most of the time. If anyone gives this a try, would be happy to hear some feedback.

---

**Update 05 Oct:**

- Added C-style API for easy use in other projects/languages
- [Initial support for Android](https://github.com/ggerganov/whisper.cpp/issues/18) - someone was able to even run the `large` model on Samsung A52!
- [Initial support for Raspberry Pi](https://github.com/ggerganov/whisper.cpp/issues/7)
- Added basic streaming example for real-time audio transcription from the mic using continuous inference with fixed audio step:

https://user-images.githubusercontent.com/1991296/193465125-c163d304-64f6-4f5d-83e5-72239c9a203e.mp4

---

**Update 30 Sep:**

Reduced memory usage even more:

| Model  | Disk   | Mem     |
| ---    | ---    | ---     |
| tiny   |  75 MB | ~240 MB |
| base   | 142 MB | ~380 MB |
| small  | 466 MB | ~970 MB |
| medium | 1.5 GB | ~2.5 GB |
| large  | 2.9 GB | ~4.6 GB |

Here is a real-world example of running the implementation to transcribe a [1h 30min video](https://www.youtube.com/watch?v=P6UKhR0T6cs) of John Carmack, on a MacBook M1 Pro (CPU only):

| Model | Time to transcribe | Result |
| ---      | ---                           | ---       |
| tiny.en | 116 sec | [carmack.wav-tiny.en.txt](https://github.com/openai/whisper/files/9686673/carmack.wav-tiny.en.txt) |
| small.en | 563 sec | [carmack.wav-small.en.txt](https://github.com/openai/whisper/files/9686680/carmack.wav-small.en.txt) |

---

**Update 29 Sep:**

Added transcription timestamps. 

---

**Update 28 Sep:**

Here is the current memory usage for the different models, using the latest implementation (Flash Attention + Flash Forward + 16-bit key/value memory):

| Model | Disk | Mem |
| ---   | --- | --- |
| tiny | 75 MB | ~460 MB |
| base | 142 MB | ~620 MB |
| small | 466 MB | ~1.3 GB |
| medium | 1.5 GB | ~2.8 GB |
| large | 2.9 GB | ~4.9 GB |

**Comment by WilliamTambellini** [comment]
Hi @ggerganov 
Congrats.
Have you tried/thought about exporting to onnx ?

**Reply by ggerganov** [reply]
Hi, I'm not familiar with onnx - need to read about it and see if it makes sense to port my implementation to it.

**Comment by FurkanGozukara** [comment]
C# wrapper would be nice :)

**Reply by ggerganov** [reply]
I will probably provide a C wrapper at some point. Just need to figure out a proper API.

**Reply by crazy4pi314** [reply]
C wrapper and then I can call from Rust <3

**Reply by ggerganov** [reply]
C-style API is now available

**Comment by dgoryeo** [comment]
Wow: " someone was able to even run the large model on Samsung A52"!.

**Reply by kevin01881** [reply]
Wait what, how??! 😮 Would love to run this on my S21 Ultra.

**Comment by Arlen22** [comment]
It took 13 minutes to do a 10 minute video on medium.en with six threads. It seems to take longer (at least to get started) if you have too many threads. Your CPU shouldn't be maxed out. Currently the whisper CPU mode doesn't even start transcribing for me, so I don't know how long it would take on that video. The video takes 3 minutes on my RTX 2060. Running Linux. 

After trying again for another 17 minutes with the whisper CPU mode it had only printed the first line. No idea what's up with that. So whisper.cpp definitely is faster :grin:

**Reply by ggerganov** [reply]
Thanks for the info.
Just a note that the `whisper.cpp` implementation currently only supports the greedy sampling strategy, so to make a fair comparison with PyTorch, you would need to disable the beam search when running it.

Regarding the threads:
So far I have observed that using more than 8 threads leads to worse performance. Even if the CPU has 24 or 32 cores. So I suggest to use `min(8, CPU_CORES)` number of threads.

**Reply by Arlen22** [reply]
I don't know what you mean by greedy sampliing, beam search, or what the difference is. Are you saying that beam search makes whisper.cpp faster or slower?

**Comment by erkkimon** [comment]
This is cool. I bumped into this because of trying your all in one script, which is fantastic! I have been wondering why isn't there JavaScript implementation for Whisper. I guess it should be possible to run Whisper client-side only since Whisper is running locally, but as you have been tinkering with this tech, what's your opinion? Will we be able to run Whisper in browsers client-side? @ggerganov

**Reply by ggerganov** [reply]
Hi, I have already ported the implementation to Javascript + WebAssembly and made a web-page that allows you to run the inference locally in a browser page. Checkout the [whisper.wasm](https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.wasm) example.

**Reply by erkkimon** [reply]
Wow. I will definitely check it out. You are genius. Respect. 👊

**Reply by erkkimon** [reply]
I tried it. Works fantastically!

**Reply by starsoccer** [reply]
Any plans to package this up into a nodejs module so it can be used outside of the browser?

**Reply by ggerganov** [reply]
Yes - npm package will be added soon

**Reply by Stvad** [reply]
@ggerganov I was wondering if you had a chance to publish it on npm :)

**Reply by ggerganov** [reply]
@Stvad You can keep track on the progress here: https://github.com/ggerganov/whisper.cpp/pull/260

**Comment by pneyrinck** [comment]
I noticed that token_timestamps is "EXPERIMENTAL." Are there tests to verify if working correctly?

**Comment by audioscavenger** [comment]
it does not work on cuda. not installing another 3GB of cuda drivers. why can't we use existing drivers? and what drivers are needed from cuda? what folder should be loaded in the PATH?

**Reply by phineas-pta** [reply]
u install cuda properly ? i saw your previous discussion but it seems u didnt install cuda

**Reply by audioscavenger** [reply]
that is correct. other projects are fine using the cudnn DLL included in various packages. this one requires 6GB of cuda installation. so I moved on.
appreciate you asking tho!

**Reply by phineas-pta** [reply]
this project requires compiling so it needs full fledge build tools sets, unlike other ready-to-use projects

**Reply by audioscavenger** [reply]
thanks! would you mind providing the compiled binaries for everyone to enjoy?

**Reply by phineas-pta** [reply]
u can grab from release section

**Reply by audioscavenger** [reply]
nope, only source code there

**Reply by phineas-pta** [reply]
is this a joke 😅

each release has 4 binaries already
